source/simulator/stokes_matrix_free.cc

```diff
@@ -1703,6 +1779,111 @@ namespace aspect
             level_cell_data[level].enable_newton_derivatives = false;
         }
     }
+
+    {
+      // Create active mesh tables to store the product of the pressure perturbation and
+      // the normalized gravity used in the free surface stabilization.
+      // Currently, mutilevel is not implemented yet, it may slow down the convergence.
+
+      // TODO: implement multilevel surface terms for the free surface stabilization.
+
+      active_cell_data.apply_stabilization_free_surface_faces = false;
+
+      if (sim.mesh_deformation
+          && !sim.mesh_deformation->get_free_surface_boundary_indicators().empty())
+        {
+          const double free_surface_theta =
+            sim.mesh_deformation->template get_matching_mesh_deformation_object<MeshDeformation::FreeSurface<dim>>()
+          .get_free_surface_theta();
+
+          // GMG doesn't support melt transport
+          AssertThrow(!sim.parameters.include_melt_transport, ExcNotImplemented());
+
+
+          active_cell_data.apply_stabilization_free_surface_faces = true;
+
+          const QGauss<dim-1> face_quadrature_formula (sim.parameters.stokes_velocity_degree+1);
+
+          const unsigned int n_face_q_points = face_quadrature_formula.size();
+
+          FEFaceValues<dim> fe_face_values (*sim.mapping,
+                                            sim.finite_element,
+                                            face_quadrature_formula,
+                                            update_values   |
+                                            update_gradients |
```

Comment:
I guess you need the gradients for the material model inputs? Because for everything else you don't seem to need them.

Reply:
Forgot to remove it.

Reply:
Jiaqi, when you get a question like this, it is better to address it by adding a comment in the code. This is something the next person looking at the code might ask as well.